File: LeetCode/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def createStack(self, head: ListNode) -> List[ListNode]:
        stack = []
        while head:
            stack.append(head)
            head = head.next
        return stack
    
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if headA and headB:
            stackA = self.createStack(headA)
            stackB = self.createStack(headB)
            intersection = True
            prev = None
            
            while stackA and stackB:
                topA = stackA.pop()
                topB = stackB.pop()
                if topA != topB: 
                    if intersection == False: # 공통된 노드가 없다는 의미
                        return
                    else: # prev는 공통된 노드
                        return prev
                else:
                    intersection = True
                    prev = topA
                
            return prev
        else:
            return

File: leetcode/Intersection_of_Two_Linked_Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null)
            return null;
        
        ListNode currA = headA;
        ListNode currB = headB;
        
        while (currA != currB) {
            // if currA reaches end of list, it switches to headB; otherwise, steps to the next
            currA = (currA == null) ? headB : currA.next;
            currB = (currB == null) ? headA : currB.next;
        }
        // listA = skipA + common
        // listB = skipB + common
        // currA moves skipA + common + skipB steps
        // currB moves skipB + common + skipA steps
        
        return currA;
    }
}

/**
 * Runtime: 1 ms, faster than 97.64% of Java online submissions for Intersection of Two Linked Lists.
 * Memory Usage: 41.6 MB, less than 79.22% of Java online submissions for Intersection of Two Linked Lists.
 */